pushy/src/main/java/com/turo/pushy/apns/ApnsChannelPool.java

```diff
@@ -254,19 +255,46 @@ public void operationComplete(final Future<Void> destroyFuture) throws Exception
      * @return a {@code Future} that will be completed when all resources held by this pool have been released
      */
     public Future<Void> close() {
-        return this.allChannels.close().addListener(new GenericFutureListener<Future<Void>>() {
+        final Promise<Void> closeDonePromise = new DefaultPromise<>(this.executor);
+        this.allChannels.close().addListener(new GenericFutureListener<Future<Void>>() {
             @Override
-            public void operationComplete(final Future<Void> future) throws Exception {
+            public void operationComplete(final Future<Void> future) {
                 ApnsChannelPool.this.isClosed = true;
 
-                if (ApnsChannelPool.this.channelFactory instanceof Closeable) {
-                    ((Closeable) ApnsChannelPool.this.channelFactory).close();
+                final Promise<Void> cancelFuturePromise = new DefaultPromise<>(ApnsChannelPool.this.executor);
+                final int numOfFuturesToCancel = pendingCreateChannelFutures.size();
+                if (numOfFuturesToCancel > 0) {
+                    final AtomicInteger numOfCanceledFutures = new AtomicInteger(0);
+                    for (Future<Channel> f : pendingCreateChannelFutures) {
+                        f.addListener(new GenericFutureListener<Future<? super Channel>>() {
```

Comment:
If I'm reading this right (and I may be missing something!), it seems like we're never actually _cancelling_ these futures, right? We're just waiting for them to finish on their own?

Reply:
Yes. I just think canceling these futures may cause `ApnsChannelPool.this.metricsListener.handleConnectionCreationFailed()` called but it is not a abnormal creation failure. And it looks safe for waiting for them to finish on their own, because these channel will be closed finally.

Reply:
In that case, I wonder if we'd be better-served by using a [`PromiseCombiner`](https://netty.io/4.1/api/io/netty/util/concurrent/PromiseCombiner.html) rather than maintaining our own countdown:

```java
final Promise<Void> waitForPendingCreateChannelFuturesPromise = new DefaultPromise<>(ApnsChannelPool.this.executor);
final PromiseCombiner combiner = new PromiseCombiner(ApnsChannelPool.this.executor);

for (final Future<Channel> future : pendingCreateChannelFutures) {
    combiner.add(future);
}

combiner.finish(waitForPendingCreateChannelFuturesPromise);
```

I think that will simplify things a lot and also clarify that we're just waiting for things to wrap up to control the order of operations rather than actively changing the state of pending operations. Seem legit?

Reply:
Also, please let me acknowledge that waitForPendingCreateChannelFuturesPromise is a terrible name and can certainly be refined ;)

Reply:
`PromiseCombiner` sounds cool. Let me fix it.

> Also, please let me acknowledge that `waitForPendingCreateChannelFuturesPromise` is a terrible name and can certainly be refined ;)

I think the name is OK. It maybe a little long but pretty descriptive.